File: src/nylon/builtins/StackBuiltins.java
package nylon.builtins;

import nylon.NylonException;
import nylon.NylonObject;
import nylon.builtins.objects.BuiltinFunction;
import nylon.nylonobjects.NylonDouble;
import nylon.parser.Parser;
import nylon.parser.parsers.BuiltinParser;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

public final class StackBuiltins {
    private static LinkedList<Stack<NylonObject>> stacks = new LinkedList<>();

    public static void build(ArrayList<Parser> parsers) {
        parsers.set('(', BuiltinParser.getParser(new BuiltinFunction('(') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                StackBuiltins.stacks.push((Stack<NylonObject>) stack.clone());
                stack.clear();
            }
        }));
        parsers.set(')', BuiltinParser.getParser(new BuiltinFunction(')') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                stack.clear();
                stack.addAll(StackBuiltins.stacks.pop());
            }
        }));
        parsers.set(':', BuiltinParser.getParser(new BuiltinFunction(':') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) throws NylonException {
                NylonObject object = stack.pop();
                try {
                    stack.add(object.clone());
                    stack.add(object.clone());
                } catch (CloneNotSupportedException e) {
                    throw new NylonException("Could not clone object \"" + object.getId() + "\"", this);
                }
            }
        }));
        parsers.set('À', BuiltinParser.getParser(new BuiltinFunction('À') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                NylonObject nylonObject = StackBuiltins.stacks.peek().pop();
                stack.push(nylonObject);
            }
        }));
        parsers.set('Á', BuiltinParser.getParser(new BuiltinFunction('Á') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                NylonObject nylonObject = stack.pop();
                StackBuiltins.stacks.peek().push(nylonObject);
            }
        }));
        parsers.set('ù', BuiltinParser.getParser(new BuiltinFunction('ù') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                NylonObject nylonObject = stack.pop();
                stack.insertElementAt(nylonObject, 0);
            }
        }));
        parsers.set('ú', BuiltinParser.getParser(new BuiltinFunction('ú') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                NylonObject nylonObject = stack.remove(0);
                stack.push(nylonObject);
            }
        }));
        parsers.set('Ï', BuiltinParser.getParser(new BuiltinFunction('Ï') {
            @Override
            public void applyImpl(Stack<NylonObject> stack) {
                NylonDouble nylonDouble = new NylonDouble(stack.size());
                stack.push(nylonDouble);
            }
        }));
    }
}

File: src/nylon/builtins/StackBuiltins.kt
package nylon.builtins

import nylon.NylonFunction
import nylon.NylonObject
import nylon.NylonStack
import nylon.nylonobjects.LongObject
import nylon.parser.parsers.BuiltinParserBuilder
import java.util.*

/**
 * Created by Aedan Smith.
 */

class PushStackParserBuilder : BuiltinParserBuilder(object : NylonFunction("(") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.stacks.push(Stack<NylonObject<*>>())
    }
}, '(')

class PopStackParserBuilder : BuiltinParserBuilder(object : NylonFunction(")") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.stacks.pop()
    }
}, ')')

class PullTopParserBuilder : BuiltinParserBuilder(object : NylonFunction("À") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.add(stack.stacks[stack.stacks.size - 2].pop())
    }
}, 'À')

class PushTopParserBuilder : BuiltinParserBuilder(object : NylonFunction("Á") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.stacks[stack.stacks.size - 2].add(stack.pop())
    }
}, 'Á')

class CloneParserBuilder : BuiltinParserBuilder(object : NylonFunction(":") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.push(stack.peek().clone())
    }
}, ':')

class MoveToBottomParserBuilder : BuiltinParserBuilder(object : NylonFunction("ù") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.add(0, stack.pop())
    }
}, 'ù')

class MoveToTopParserBuilder : BuiltinParserBuilder(object : NylonFunction("ú") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.push(stack.removeAt(0))
    }
}, 'ú')

class LengthOfStackParserBuilder : BuiltinParserBuilder(object : NylonFunction("Ï") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.push(LongObject(stack.getSize().toLong()))
    }
}, 'Ï')
